polkadot/runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -1168,3 +1202,84 @@ fn filter_backed_statements_from_disabled_validators<T: shared::Config + schedul
 	filtered || backed_len_before != backed_candidates.len()
 }
+
+/// Filter out all candidates that have multiple cores assigned and no
+/// `CoreIndex` injected.
+/// Returns whether or not we dropped any candidates.
+fn filter_elastic_scaling_candidates<
+	T: configuration::Config + scheduler::Config + inclusion::Config,
+>(
+	allowed_relay_parents: &AllowedRelayParentsTracker<T::Hash, BlockNumberFor<T>>,
+	core_index_enabled: bool,
+	candidates: &mut Vec<BackedCandidate<T::Hash>>,
+) -> bool {
+	// Count how many scheduled cores each paraid has.
+	let mut cores_per_parachain: BTreeMap<ParaId, usize> = BTreeMap::new();
+
+	for (_, para_id) in <scheduler::Pallet<T>>::scheduled_paras() {
+		*cores_per_parachain.entry(para_id).or_default() += 1;
+	}
+
+	let prev_count = candidates.len();
+	// We keep a candidate if the parachain has only one core assigned or if
+	// a core index is provided by block author.
+	candidates.retain(|candidate| {
+		*cores_per_parachain.get(&candidate.descriptor().para_id).unwrap_or(&0) <= 1 ||
+			has_core_index::<T>(allowed_relay_parents, candidate, core_index_enabled)
+	});
+
+	prev_count != candidates.len()
+}
+
+// Returns `true` if the candidate contains a valid injected `CoreIndex`.
+fn has_core_index<T: configuration::Config + scheduler::Config + inclusion::Config>(
+	allowed_relay_parents: &AllowedRelayParentsTracker<T::Hash, BlockNumberFor<T>>,
+	candidate: &BackedCandidate<T::Hash>,
+	core_index_enabled: bool,
+) -> bool {
+	// After stripping the 8 bit extensions, the `validator_indices` field length is expected
+	// to be equal to backing group size. If these don't match, the `CoreIndex` is badly encoded,
+	// or not supported.
+	let (validator_indices, maybe_core_idx) =
+		candidate.validator_indices_and_core_index(core_index_enabled);
+
+	let Some(core_idx) = maybe_core_idx else { return false };
+
+	let relay_parent_block_number =
+		match allowed_relay_parents.acquire_info(candidate.descriptor().relay_parent, None) {
+			Some((_, block_num)) => block_num,
+			None => {
+				log::debug!(
+					target: LOG_TARGET,
+					"Relay parent {:?} for candidate {:?} is not in the allowed relay parents. Dropping the candidate.",
+					candidate.descriptor().relay_parent,
+					candidate.candidate().hash(),
+				);
+				return false
+			},
+		};
+
+	// Get the backing group of the candidate backed at `core_idx`.
+	let group_idx = match <scheduler::Pallet<T>>::group_assigned_to_core(
+		core_idx,
+		relay_parent_block_number + One::one(),
+	) {
+		Some(group_idx) => group_idx,
+		None => {
+			log::debug!(
+				target: LOG_TARGET,
+				"Can't get the group index for core idx {:?}. Dropping the candidate {:?}.",
+				core_idx,
+				candidate.candidate().hash(),
+			);
+			return false
+		},
+	};
+
+	let group_validators = match <scheduler::Pallet<T>>::group_validators(group_idx) {
+		Some(validators) => validators,
+		None => return false,
+	};
+
+	group_validators.len() == validator_indices.len()
```

Comment:
yeah, we're already doing the check here: https://github.com/paritytech/polkadot-sdk/blob/fee810a5ea177f95ddf13ea51270b84f1e117bae/polkadot/primitives/src/v6/mod.rs#L755

Reply:
> If this is not true, then the check is redundant here and should be removed.

I wonder if it's safe to remove it. After we call filter_elastic_scaling_candidates(), we filter out backing votes of disabled validators (before doing the check I linked above). So if ElasticScalingMVP is enabled but a validator was not updated, we'd maybe wrongfully filter disabled votes because we wouldn't realise that the core index was not actually encoded (only if the backing group size is larger than 8).

So we'd calculate a wrong core index for the candidate and possibly wrongfully remove the backing of the candidate which had not encoded the core index after ElasticScalingMVP was enabled.

Considering that this would require a backing group size larger than 8 (which we don't have on any network AFAIK) and that I think it only enables the auto-censorship of unupgraded nodes, I'd say it's fine and we could remove this check here. WDYT?

Reply:
let's keep as is then.